### PR TITLE
feat: support resume urls

### PR DESCRIPTION
**Pull request by Deivu**
**Please describe the changes this PR makes and why it should be merged:**
Resume urls was not added in v14 despite on changelogs saying it is, was only added on `@discordjs/ws`

**Status and versioning classification:**
- Code changes have been tested against the Discord API, or there are no code changes
- I know how to update typings and have done so, or typings don't need updating

### CODE REVIEW

**Review by almeidx — CHANGES_REQUESTED** [review]

**Review by didinele — APPROVED** [review]

**Review by didinele — CHANGES_REQUESTED** [review]
Actually, in hindsight, you aren't supposed to use the resume URL in case of a full reconnect. You should be clearing it in `destroy` if `reset` is true.

File: packages/discord.js/src/client/websocket/WebSocketShard.js [hunk 1]
```diff
@@ -64,6 +64,13 @@ class WebSocketShard extends EventEmitter {
      */
     this.sessionId = null;
 
+    /**
+     * The resume url for this shard
+     * @type {?string}
+     * @private
+     */
+    this.resumeUrl = null;
```
Comment (Syjalo) [inline packages/discord.js/src/client/websocket/WebSocketShard.js:72]:
```suggestion
    this.resumeURL = null;
```

File: packages/discord.js/src/client/websocket/WebSocketShard.js [hunk 2]
```diff
@@ -193,12 +200,14 @@ class WebSocketShard extends EventEmitter {
...
    */
   connect() {
-    const { gateway, client } = this.manager;
+    const { client } = this.manager;
 
     if (this.connection?.readyState === WebSocket.OPEN && this.status === Status.Ready) {
       return Promise.resolve();
     }
 
+    const gateway = this.resumeUrl ?? this.manager.gateway;
```
Comment (Syjalo) [inline packages/discord.js/src/client/websocket/WebSocketShard.js:209]:
```suggestion
    const gateway = this.resumeURL ?? this.manager.gateway;
```

File: packages/discord.js/src/client/websocket/WebSocketShard.js [hunk 3]
```diff
@@ -416,9 +425,10 @@ class WebSocketShard extends EventEmitter {
         this.emit(WebSocketShardEvents.Ready);
 
         this.sessionId = packet.d.session_id;
+        this.resumeUrl = packet.d.resume_gateway_url;
```
Comment (Syjalo) [inline packages/discord.js/src/client/websocket/WebSocketShard.js:428]:
```suggestion
        this.resumeURL = packet.d.resume_gateway_url;
```

File: packages/discord.js/src/client/websocket/WebSocketShard.js [hunk 4]
```diff
@@ -416,9 +425,10 @@ class WebSocketShard extends EventEmitter {
...
         this.expectedGuilds = new Set(packet.d.guilds.map(d => d.id));
         this.status = Status.WaitingForGuilds;
-        this.debug(`[READY] Session ${this.sessionId}.`);
+        this.debug(`[READY] Session ${this.sessionId} | Resume url ${this.resumeUrl}.`);
```
Comment (Syjalo) [inline packages/discord.js/src/client/websocket/WebSocketShard.js:431]:
```suggestion
        this.debug(`[READY] Session ${this.sessionId} | Resume url ${this.resumeURL}.`);
```

File: packages/discord.js/typings/index.d.ts [hunk 5]
```diff
@@ -2994,6 +2994,7 @@ export class WebSocketShard extends EventEmitter {
   private sequence: number;
   private closeSequence: number;
   private sessionId: string | null;
+  private resumeUrl: string | null;
```
Comment (Syjalo) [inline packages/discord.js/typings/index.d.ts:2997]:
```suggestion
  private resumeURL: string | null;
```

**Review by Syjalo — CHANGES_REQUESTED** [review]

**Comment by Deivu** [comment]
Discord.JS uses pascal case so why we need url to be all capital?

**Comment by didinele** [comment]
> Discord.JS uses pascal case so why we need url to be all capital?

We always uppercase abbreviations, e.g. `REST`, or https://discord.js.org/#/docs/discord.js/main/typedef/WebhookClientDataURL - you can find plenty of examples of this.

**Comment by Deivu** [comment]
Ok, then applied the changes you requested

**Review by didinele — APPROVED** [review]

**Review by SpaceEEC — APPROVED** [review]

**Review by kyranet — APPROVED** [review]

File: packages/discord.js/src/client/websocket/WebSocketShard.js [hunk 6]
```diff
@@ -64,6 +64,13 @@ class WebSocketShard extends EventEmitter {
      */
     this.sessionId = null;
 
+    /**
+     * The resume url for this shard
```
Comment (Jiralite) [inline packages/discord.js/src/client/websocket/WebSocketShard.js:68]:
```suggestion
     * The resume URL for this shard
```

File: packages/discord.js/src/client/websocket/WebSocketShard.js [hunk 7]
```diff
@@ -842,10 +852,11 @@ class WebSocketShard extends EventEmitter {
     // Step 4: Cache the old sequence (use to attempt a resume)
     if (this.sequence !== -1) this.closeSequence = this.sequence;
 
-    // Step 5: Reset the sequence and session id if requested
+    // Step 5: Reset the sequence, resume url and session id if requested
```
Comment (Jiralite) [inline packages/discord.js/src/client/websocket/WebSocketShard.js:855]:
```suggestion
    // Step 5: Reset the sequence, resume URL and session id if requested
```

**Review by iCrawl — APPROVED** [review]

**Pull request merged**